apisix/utils/batch-processor.lua

```diff
@@ -161,7 +161,7 @@ function batch_processor:push(entry)
     local entries = self.entry_buffer.entries
     table.insert(entries, entry)
     -- add batch metric for every route
-    if batch_metrics  then
+    if batch_metrics and self.name and self.route_id and self.server_addr then
```

Comment:
We could provide a default `''` for the route_id, right?

> Sometimes we need to judge whether matching route accroding to toute_id.

There is no need to judge with route_id for `error-log-logger`.

Reply:
Do you mean provider `''` for route_id on
https://github.com/apache/apisix/blob/ddb9cd28bf33799652d252df0546aa0832933bcf/apisix/utils/batch-processor.lua#L138
We can do it, but I think it doesn't change the conditional judgment here. How about you ?

Reply:
Yes. So that we can remove repeated conditional judgment.

Reply:
If removing the repeated conditional judgment, I think should provider a default value `''` for `server_addr`, error-log-logger doesn't have the `server_addr` too.

Reply:
We can use `ctx.var.server_addr` for it.

Reply:
I add two line `local ngx = ngx` and  `core.log.error("ngx.var.server_addr",ngx.var.server_addr)`  on batch-processor.lua  but got an error. It seems it deson't use `ngx` on log phase.
```
2022/01/11 22:23:46 [error] 21652#21652: *217 lua user thread aborted: runtime error: /usr/local/apisix/apisix/utils/batch-processor.lua:138: API disabled in the current context
stack traceback:
coroutine 0:
        [C]: in function 'error'
        /usr/local/openresty/lualib/resty/core/var.lua:103: in function '__index'
        /usr/local/apisix/apisix/utils/batch-processor.lua:138: in function 'new'
        /usr/local/apisix/apisix/plugins/error-log-logger.lua:316: in function </usr/local/apisix/apisix/plugins/error-log-logger.lua:250>
coroutine 1:
        [C]: in function 'thread_spawn'
        /usr/local/apisix/apisix/timers.lua:41: in function </usr/local/apisix/apisix/timers.lua:32>
        [C]: in function 'pcall'
        /usr/local/apisix/apisix/core/timer.lua:35: in function </usr/local/apisix/apisix/core/timer.lua:31>
        [C]: in function 'pcall'
        /usr/local/apisix/apisix/core/timer.lua:58: in function </usr/local/apisix/apisix/core/timer.lua:51>, context: ngx.timer
2022/01/11 22:23:46 [error] 21649#21649: *218 lua user thread aborted: runtime error: /usr/local/apisix/apisix/utils/batch-processor.lua:138: API disabled in the current context
stack traceback:
coroutine 0:
        [C]: in function 'error'
        /usr/local/openresty/lualib/resty/core/var.lua:103: in function '__index'
        /usr/local/apisix/apisix/utils/batch-processor.lua:138: in function 'new'
        /usr/local/apisix/apisix/plugins/error-log-logger.lua:316: in function </usr/local/apisix/apisix/plugins/error-log-logger.lua:250>
coroutine 1:
        [C]: in function 'thread_spawn'
        /usr/local/apisix/apisix/timers.lua:41: in function </usr/local/apisix/apisix/timers.lua:32>
        [C]: in function 'pcall'
        /usr/local/apisix/apisix/core/timer.lua:35: in function </usr/local/apisix/apisix/core/timer.lua:31>
        [C]: in function 'pcall'
        /usr/local/apisix/apisix/core/timer.lua:58: in function </usr/local/apisix/apisix/core/timer.lua:51>, context: ngx.timer
2022/01/11 22:23:46 [error] 21652#21652: *217 [lua] timers.lua:54: failed to wait threads: API disabled in the current context, context: ngx.timer
2022/01/11 22:23:46 [error] 21649#21649: *218 [lua] timers.lua:54: failed to wait threads: API disabled in the current context, context: ngx.timer
2022/01/11 22:23:46 [error] 21648#21648: *223 lua user thread aborted: runtime error: /usr/local/apisix/apisix/utils/batch-processor.lua:138: API disabled in the current context
stack traceback:
coroutine 0:
        [C]: in function 'error'
        /usr/local/openresty/lualib/resty/core/var.lua:103: in function '__index'
        /usr/local/apisix/apisix/utils/batch-processor.lua:138: in function 'new'
        /usr/local/apisix/apisix/plugins/error-log-logger.lua:316: in function </usr/local/apisix/apisix/plugins/error-log-logger.lua:250>
coroutine 1:
        [C]: in function 'thread_spawn'
        /usr/local/apisix/apisix/timers.lua:41: in function </usr/local/apisix/apisix/timers.lua:32>
        [C]: in function 'pcall'
        /usr/local/apisix/apisix/core/timer.lua:35: in function </usr/local/apisix/apisix/core/timer.lua:31>
        [C]: in function 'pcall'
        /usr/local/apisix/apisix/core/timer.lua:58: in function </usr/local/apisix/apisix/core/timer.lua:51>, context: ngx.timer
2022/01/11 22:23:46 [error] 21648#21648: *223 [lua] timers.lua:54: failed to wait threads: API disabled in the current context, context: ngx.timer



```

Reply:
I see. We can't use `ngx.var` in the timer. What about refactoring `self.name and self.route_id and self.server_addr` into a function? Would be better if we don't need to repeat the complex conditions.

Reply:
okay